File: conversions/DigitsToWords.scala
import scala.annotation.tailrec

object DigitsToWords {

  private val Wordings = Array(
    "zero",
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
  )

  // Converts digits to words, for example, 1985 will be converted to "one nine eight five"
  def convert(n: Int): String = {
    @tailrec
    def convertAccumulatingResult(n: Int, acc: Seq[String]): String = {
      if (n > 0) {
        convertAccumulatingResult(n / 10, Wordings(n % 10) +: acc)
      } else {
        acc.mkString(" ")
      }
    }

    convertAccumulatingResult(n, Seq.empty)
  }
}

File: convert/digits-to-words/DigitsToWords.hs
-- Converts digits to the words, for example, 1985 will be displayed as "one nine eight five".

import Data.Char (isSpace)

type Wording = [Char]

wordings :: [Wording]
wordings =  ["zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

convert :: Int -> Wording
convert n
  | n == 0 = head wordings
  | otherwise =
    let s = match $ abs n
    in if n < 0 then "minus" ++ s else dropWhile isSpace s
  where
    match 0 = ""
    match x = match (x `div` 10) ++ " " ++ wordings !! (x `mod` 10)
